Adds in-mem `PolicyReportStore` implementation (#107)

nnelas opened the pull request:
## Description

<!-- Please provide the link to the GitHub issue you are addressing -->
Fix #107

<!-- Please provide the link to the documentation related to your change, if applicable -->
<!-- [Documentation](https://<insert your url>) -->

As specified in the original issue, this commit:
- Create a store Go interface that satisfies the current store implementation at `report/store.go`
- Refactor current `report/store.go` into `report/store_kubernetes.go`, as an implementation of the store interface
- Adds a new memory store with 2 maps, one with `ClusterPolicyReports` and another one with `PolicyReports`
- Provide an optional cli flag, `--store TYPE`, with possible types memory, kubernetes. Defaults to --store kubernetes for ease on first Kubewarden installations.

Besides that, this commit also:
- Changes `printJSON` to `outputScan` to make it more aligned with what the input params is
- `outputScan` only outputs at the end of the flow, to prevent having a lot of logs whenever someone scans the entire cluster (we were outputting the entire store on each namespace loop)

It also seems that running `make fmt` introduced a couple of unwanted changes that I'm happy to revert to keep this commit smaller. Finally, I also had to do a couple of tweaks to make golangci lint happy.

## Test

<!-- Please provides a short description about how to test your pullrequest -->
To test this pull request, you can run the following commands:

```shell
make unit-tests
```

## Additional Information

### Tradeoff

<!-- Please describe, if any, the tradeoffs that you found acceptable in this pull request -->

### Potential improvement

<!-- Please describe, if any, potential improvement that you are envisioning -->
One thing that we can also do in the future is to parse (in some way) the `toJSON()` result and instead of logging the entire report in a single log block, we could have a log line per violation. What do you think?

Also, I'll be out-of-office for the rest of the week, but I'll return next Monday. Looking forward to seeing your comments! ❤️

jvanz commented on internal/report/store.go:
```diff
@@ -1,291 +1,45 @@
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	polReport "sigs.k8s.io/wg-policy-prototypes/policy-report/pkg/api/wgpolicyk8s.io/v1alpha2"
+const (
+	KUBERNETES string = "KUBERNETES"
+	MEMORY     string = "MEMORY"
```
```suggestion
	KUBERNETES string = "kubernetes"
	MEMORY     string = "memory"
```

nnelas replied:
changed! :)

jvanz reviewed (changes requested):
I could not finish testing the PR yet. But I'll continue soon.

jvanz commented on internal/report/store_memory.go:
```diff
@@ -0,0 +1,198 @@
+		return getErr
+	}
+	// Update existing Policy Report
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// get the latest report version to be updated
+		latestReport, err := s.GetPolicyReport(report.GetNamespace())
+		if err != nil {
+			return err
+		}
+		latestReport.Summary = report.Summary
+		latestReport.Results = report.Results
+		return s.UpdatePolicyReport(&latestReport)
+	})
```
I think you can remove the `RetryOnConflict`. This is use in K8s report because of the conflict it can happen when updating the report in the backend. Here, I don't expect that to happen so often. Therefore, checking the `GetPolicyReport` error once is enough.

nnelas replied:
that makes sense, thank you for noticing that! changed for `SavePolicyReport` and [SaveClusterPolicyReport](https://github.com/kubewarden/audit-scanner/pull/150#discussion_r1417270463)

jvanz commented on internal/report/store_memory.go:
```diff
@@ -0,0 +1,198 @@
+		return getErr
+	}
+	// Update existing Policy Report
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
I think you can remove the `RetryOnConflict`. This is use in K8s report because of the conflict it can happen when updating the report in the backend. Here, I don't expect that to happen so often. Therefore, checking the `GetPolicyReport` error once is enough.

viccuad commented on cmd/root.go:
```diff
@@ -27,63 +29,92 @@ type Scanner interface {
+			return err
+		}
+
+		if outputScan {
```
I think it should stay. The in memory store should be there just for testing. We implement the interface in the memory store or for example a future SQL backend.

nnelas replied:
my apologies @viccuad, but I didn't understood what you meant with
> I think we should have this logic to do ToJSON and print in cmd/root.go, and it should be on scanner.go.

viccuad replied:
I meant that I would like the block `if outputScan { scanner.ReportStore.ToJSON() }` to stay in `scanner/scanner.go` as it was before. This puts all the logic in the internal libs, and leaves `cmd/*` thinner, just dealing with flags and making the big calls. It will simplify future testing and refactoring too.

nnelas replied:
I get your point now, thank you for your explanation @viccuad ! one thing that it's important to mention is that this way we will output all store data (either Kubernetes or Memory) multiple times per namespace which is also not ideal

viccuad commented on README.md:
```diff
@@ -10,10 +10,13 @@
+Instead of relying on `PolicyReport` objects, one can also configure Audit scanner to
+save all this information in-memory only, by specifying `--store MEMORY`.
+
 # Deployment
```
(opening thread here randomly)

> One thing that we can also do in the future is to parse (in some way) the toJSON() result and instead of logging the entire report in a single log block, we could have a log line per violation. What do you think?

At that point, I think it would be better  to expose the reports with proper structured logging.

nnelas replied:
yeap, completely agree with that. but perhaps it can be another feature request with another PR, or do you think it's feasible to include that implementation under this one?

flavio replied:
I too think it would be better to get this done with another PR

viccuad commented on cmd/root.go:
```diff
@@ -27,63 +29,92 @@ type Scanner interface {
+			return err
+		}
+
+		policyReportStore, err := getPolicyReportStore(storeType)
```
We should be creating the store in scanner.go as we were doing before, and not in `cmd/root.go`. Ideally, `cmd/root.go` should be only used for reading the cmd flags, and calling an audit-scanner lnternal library with all the logic. Right now `cmd/root.go` already does more than it should (create policiesFetcher, resourcesFetcher, etc) but that should change in the future.

Refactoring into an internal lib would simplify testing and mocks, yet it is out of the scope for this PR. But that doesn't mean we shouldn't keep piling up things in `cmd/root.go`.

nnelas replied:
I completely agree with you, in this case, I went to the easy path and decided to add another dependency to `cmd/root.go` when shouldn't. I've refactored this bit with your suggestions. there's still `outputScan` there that I'm happy to move it elsewhere after https://github.com/kubewarden/audit-scanner/pull/150#discussion_r1417210469

viccuad commented on internal/report/store.go:
```diff
@@ -1,291 +1,45 @@
-	return nil
-}
+	// UpdateClusterPolicyReport updates a ClusterPolicyReport or PolicyReport.
+	// ns argument is used in case of namespaced PolicyReport
```
there is no ns argument, nor this function updates a PolicyReport

nnelas replied:
although I should've been more careful when moving the original godocs, this method will be removed with https://github.com/kubewarden/audit-scanner/pull/150#discussion_r1419164395

viccuad commented on internal/report/store_memory.go:
```diff
@@ -0,0 +1,198 @@
+	// cprCache is a map between a name and a ClusterPolicyReport
+	cprCache map[string]ClusterPolicyReport
+
+	mutex sync.RWMutex
```
I think we should have them, we can't ensure it's write and read is concurrent safe.

viccuad replied:
True (yet they don't hurt now that they are there, so for me it's not a merge blocker to remove them).

nnelas replied:
I think I was just suffering from anticipation, I will remove those locks since we don't need them yet

viccuad replied:
I'm ok with them being removed for now :).

viccuad commented on internal/report/store.go:
```diff
@@ -1,291 +1,45 @@
-}
+	// UpdateClusterPolicyReport updates a ClusterPolicyReport or PolicyReport.
+	// ns argument is used in case of namespaced PolicyReport
+	UpdateClusterPolicyReport(report *ClusterPolicyReport) error
```
Since only `SaveClusterPolicyReport` and `SavePolicyReport` are used from outside of the package, we should not have `UpdateClusterPolicyReport` and `UpdatePolicyReport` as part of the interface, and they should be private on the implemented packages.

nnelas replied:
to be honest, I had mixed feelings about this one, I did notice that these 2 methods were not being used outside of the package but I didn't want to change the interface since you could've left them public for future work. in this case, I'll remove them from the interface

nnelas replied:
also, with that in mind, does it make sense to remove both `RemovePolicyReport` and `RemoveAllNamespacedPolicyReports` given that we currently have no usage for them?

jvanz replied:
Yeah, if we are not using, I'm fine removing these methods. I think that  code no in use should not be in the code base.

nnelas replied:
completely agree @jvanz , I'll remove those `Remove*` methods

viccuad commented on internal/report/store_memory_test.go:
```diff
@@ -0,0 +1,173 @@
+package report_test
+
+import (
+	"github.com/stretchr/testify/require"
```
I would prefer not to add a (dev) dependency if it isn't needed, maybe we can do something else besides `require.NoError()`.

nnelas replied:
make sense, I ended up using a plain nil-check for errors and removed `require` as a dependency

viccuad replied:
Many thanks!

nnelas commented on internal/report/store_memory.go:
```diff
@@ -0,0 +1,198 @@
+	s.prCache[report.GetNamespace()] = *report
+
+	summary, _ := report.GetSummaryJSON()
+	log.Info().
```
agreed! I've pushed a new commit that just align these log levels and their messages between `PolicyReports`, `ClusterPolicyReports` and policy stores

nnelas commented on internal/policies/fetcher.go:
```diff
@@ -194,7 +194,7 @@ func (f *Fetcher) getAdmissionPolicies(namespace string) ([]policiesv1.Admission
 }
 
-func newClient() (client.Client, error) { //nolint:ireturn
+func newClient() (client.Client, error) { //nolint:ireturn // no specific client type
```
to be fair, this was a quick and dirty "solution" to try to keep `ireturn` lint happy. for some reason, it's a bit flaky as sometimes `nolintlint` complains that that comment is not needed but it's not consistent :/

nnelas replied:
thank you so much for your tip! It did the trick 🎉

nnelas commented on internal/report/store.go:
```diff
@@ -1,291 +1,45 @@
-	return nil
-}
+// PolicyReportStore caches the latest version of PolicyReports
+// There is 2 different stores available:
```
my goal here was to have an easy way for any client of PolicyReportStore API to know which implementations are available and what's their goal, but you're right, doesn't make sense to have an API knowing about it's implementations

nnelas commented:
hallo @jvanz and @viccuad ! first of all, thank you so much for your suggestions/comments, really helpful to keep the codebase consistent and aligned with your standards ❤️ 

I think I addressed all your topics, except for https://github.com/kubewarden/audit-scanner/pull/150#discussion_r1417210469 and https://github.com/kubewarden/audit-scanner/pull/150#discussion_r1417221641 since I need further help/clarification for them

jvanz commented on internal/policies/fetcher_test.go:
```diff
@@ -357,20 +361,22 @@ var nsDefault = v1.Namespace{
 }
 
-func mockClient(initObjs ...k8sClient.Object) k8sClient.Client { //nolint:ireturn
+func mockClient(initObjs ...k8sClient.Object) k8sClient.Client { //nolint:ireturn // no specific client type
```
Again, try the trick about the linter that I've shared in another comment

jvanz reviewed (changes requested):
Thanks @nnelas! I think we are almost merging this PR :heart: 

I just have some small requests about the linter. Besides that, I would say we can wait for another @viccuad review and merge if he approved.

jvanz commented:
Furthermore, @nnelas can you rebase and fix the conflicts in the `go.sum` file? :grin:

flavio commented on internal/report/store_memory.go:
```diff
@@ -0,0 +1,172 @@
+	cprCache map[string]ClusterPolicyReport
+}
+
+func NewMemoryPolicyReportStore() (*MemoryPolicyReportStore, error) {
```
Since it never fails, we could omit `error` from the return type

nnelas replied:
I had mixed feelings about this one, I was thinking of keeping all constructors for all `PolicyReportStore` implementations coherent, but I'm now thinking that doesn't make sense because we'll have different dependencies for each store. I'll remove the `error`

flavio reviewed (commented):
Great contribution, thanks a lot @nnelas!

jvanz reviewed (approved):
LGTM. @kubewarden/kubewarden-developers remember to squash the commits when merging this PR!

viccuad commented on go.mod:
```diff
@@ -8,14 +8,15 @@ require (
 	k8s.io/apimachinery v0.29.0
 	k8s.io/client-go v0.29.0
 	sigs.k8s.io/controller-runtime v0.16.3
-	sigs.k8s.io/wg-policy-prototypes v0.0.0-20230505033312-51c21979086a
```
This represents a bump in the wg-policy-prototypes that changes the field of the crds. See the comparison here:
https://github.com/kubernetes-sigs/wg-policy-prototypes/compare/51c21979086a654f60d5d65a95f9681e954816ac..22764d64b0c3f79d2293b67000b4d9ebca197623

Particularly the file `policy-report/crd/v1beta1/wgpolicyk8s.io_clusterpolicyreports.yaml`. There's only 2 new fields to the CRDs, and no deletions, but that means that PolicyReports created with this new library version may not be backwards compatible.

Thing is, https://github.com/kubernetes-sigs/wg-policy-prototypes is not correctly tagged nor has proper releases, and other community consumers keep just developing the future v1beta1 (listed as [WIP](https://github.com/kubernetes-sigs/wg-policy-prototypes/tree/master/policy-report#policy-report)), so we need to be careful.

I would like to postpone this bump.

nnelas replied:
oh, I see! that makes sense, let me revert that change in a bit

viccuad commented on internal/report/store_kubernetes.go:
(for future PRs if there's need to move files around (like store.go -> store_kubernetes.go) I would prefer the usage of `git mv` to create a distinct commit for it (easier to review, and maintains original authorship).

nnelas replied:
didn't know about that! thanks for the TIL 🙇

viccuad reviewed (commented):
Looks good! Many thanks for dealing with the changes!

The only blocker to me is the bump on the `sigs.k8s.io/wg-policy-prototypes` library, I would like to postpone that.

nnelas commented:
> Looks good! Many thanks for dealing with the changes!
> 
> The only blocker to me is the bump on the `sigs.k8s.io/wg-policy-prototypes` library, I would like to postpone that.

hallo @viccuad ! thank you so much for your review 🙇 I've reverted that upgrade under https://github.com/kubewarden/audit-scanner/pull/150/commits/10c4eee528da06b71fc00241f1cb3836b705e0d5 so I think I've addressed all your suggestions. in the meantime, I'll create a PR in https://github.com/kubewarden/helm-charts to use this new `--store` flag

viccuad reviewed (approved):
LGTM, thanks!

viccuad commented:
golangci-lint was again cryptically failing, and passing locally. Ended up deleting its cache under https://github.com/kubewarden/audit-scanner/actions/caches.

GH says merging is blocked because all commits must be signed, yet [DCO reports all are signed](https://github.com/kubewarden/audit-scanner/pull/150/checks?check_run_id=19681233535) Oo. Maybe it got confused with unsigned merge commits from gh.

@flavio could you merge this please?

flavio commented:
Squashed and merged. Thanks for your contribution @nnelas!
